lib/postgrex/notifications.ex

```diff
@@ -115,60 +120,72 @@ defmodule Postgrex.Notifications do
     # If this is the first listener for the given channel,
     # we need to actually issue the LISTEN query.
     if Map.size(s.listener_channels[channel]) == 1 do
-      listener_query("LISTEN \"#{channel}\"", {:ok, ref}, from, s)
+      listener_query("LISTEN \"#{channel}\"", {:ok, ref}, from, s) |> include_idle_timeout
     else
-      {:reply, {:ok, ref}, s}
+      {:reply, {:ok, ref}, s} |> include_idle_timeout
     end
   end
 
   def handle_call({:unlisten, ref}, from, s) do
     case Map.fetch(s.listeners, ref) do
       :error ->
-        {:reply, {:error, %ArgumentError{}}, s}
+        {:reply, {:error, %ArgumentError{}}, s} |> include_idle_timeout
 
       {:ok, {channel, _pid}} ->
         Process.demonitor(ref, [:flush])
         s = remove_monitored_listener(s, ref, channel)
 
         if Map.size(s.listener_channels[channel]) == 0 do
           s = update_in(s.listener_channels, &Map.delete(&1, channel))
-          listener_query("UNLISTEN \"#{channel}\"", :ok, from, s)
+          listener_query("UNLISTEN \"#{channel}\"", :ok, from, s) |> include_idle_timeout
         else
-          {:reply, :ok, s}
+          {:reply, :ok, s} |> include_idle_timeout
         end
     end
   end
 
   def handle_info({:DOWN, ref, :process, _, _}, s) do
     case Map.fetch(s.listeners, ref) do
       :error ->
-        {:noreply, s}
+        {:noreply, s} |> include_idle_timeout
 
       {:ok, {channel, _pid}} ->
         s = remove_monitored_listener(s, ref, channel)
 
         if Map.size(s.listener_channels[channel]) == 0 do
           s = update_in(s.listener_channels, &Map.delete(&1, channel))
-          listener_query("UNLISTEN \"#{channel}\"", :ok, nil, s)
+          listener_query("UNLISTEN \"#{channel}\"", :ok, nil, s) |> include_idle_timeout
         else
-          {:noreply, s}
+          {:noreply, s} |> include_idle_timeout
         end
     end
   end
 
+  def handle_info(:timeout, %{protocol: protocol} = state) do
+    {:ok, protocol} = Protocol.ping(protocol)
```

Comment:
Regarding the ok/error, you should do exactly like in this case here: https://github.com/elixir-ecto/postgrex/pull/388/files#diff-121fce3140827f25d3c6efc58a82dac7R173